LIP-36: NEST (#116)
lidofinance/lido-improvement-proposals: LIPS/lip-36.md

---
lip: 36
title: NEST — Automated LDO Buyback and Liquidity Provisioning System
status: Proposed
author: Vasiliy Shapovalov, Vitaly Galaichuk, Jen Kopytina, Alexander Belokon, adcv
discussions-to: https://research.lido.fi/t/liquid-buybacks-nest-execution-with-ldo-wsteth-liquidity/10894

> created: 2026-04-20

Reviewer: pls add `updated:`

Reviewer 2: fixed

> created: 2026-04-20

Reviewer: pls add `updated:`

---

## Simple Summary

> NEST is a deterministic onchain system that creates a direct rule-based economic link between Lido DAO success and LDO. It enables permissionless conversion of an excess share of staking revenue into LDO via CoW Swap and pair it with wstETH as DAO-owned liquidity in a Curve LDO/wstETH pool. NEST uses explicit spending caps and formulas to determine available surplus. It must be explicitly funded in order to be operational. At launch, it will track staking revenue only; the architecture is designed to accommodate additional revenue sources through future governance votes. NEST can also operate in treasury-only mode if the DAO decides to skip LP provisioning and send converted LDO directly to the treasury.

Author: Fixed in [1c7eb10](https://github.com/lidofinance/lido-improvement-proposals/pull/116/commits/1c7eb1042f21e92aa3fd5a438f07f195e688ea9b).

## Motivation

### Background

> Lido DAO generates substantial protocol revenue through staking operations. LDO token holders, however, have no direct automatic link to protocol performance. The DAO participants has discussed mechanisms to address this in the forum thread referenced above. But no formal technical proposal or agreed architecture was proposed.

Author: Fixed in [1c7eb10](https://github.com/lidofinance/lido-improvement-proposals/pull/116/commits/1c7eb1042f21e92aa3fd5a438f07f195e688ea9b).

> Lido DAO generates substantial protocol revenue through staking operations. LDO token holders, however, have no direct automatic link to protocol performance. The DAO participants have discussed mechanisms to address this in the forum thread referenced above. But no formal technical proposal or agreed architecture was proposed.

Reviewer: ```suggestion
Lido DAO generates substantial protocol revenue through staking operations. LDO token holders, however, have no direct automatic link to protocol performance. The DAO participants have discussed mechanisms to address this in the [forum thread](https://research.lido.fi/t/liquid-buybacks-nest-execution-with-ldo-wsteth-liquidity/10894) referenced above. But no formal technical proposal or agreed architecture was proposed.
```

Reviewer 2: fixed

### Problem

Without a rule-based surplus conversion mechanism, the DAO faces two structural issues:

1. Revenue above operating costs is absorbed into discretionary spend with no formal commitment to token holders.
2. Any future buyback initiative would require designing a mechanism under pressure, without calibrated parameters or governance controls already in place.

Additionally, on-chain LDO liquidity depth is shallow. Without a systematic provisioning mechanism, any buyback program risks slippage-induced execution failures as it consumes available market depth.

### Solution

> NEST encodes a governance commitment: when protocol surplus conditions are met, a constrained share of excess staking revenue is automatically converted into LDO and, in LP mode (the initial configuration proposed by this LIP), paired with wstETH and deployed as DAO-owned liquidity. The mechanism is designed to be financially safe, permissionless in normal operation, low-maintenance, and fully controllable by the DAO.

Reviewer: ```suggestion
NEST encodes a governance commitment: when protocol surplus conditions are met, a constrained share of excess staking revenue is automatically converted into LDO and, in LP mode (the initial configuration proposed by this LIP), paired with wstETH and deployed as DAO-owned liquidity. The mechanism is designed to be bounded by spending caps and governance controls, permissionless in normal operation, low-maintenance, and fully controllable by the Lido DAO.
```

Reviewer: let's avoid 'financially safe' wording please

Reviewer 2: fixed

## Specification

### System Architecture

NEST is implemented as two cooperating but independently operable subsystems — Trade Execution and Liquidity Provisioning. They utilize existing contracts like Stonks v2 and OracleRouter, as well as the governance infrastructure.

The system operates in two governance-controlled modes:

> 1. **LP mode** — acquired LDO is paired with wstETH and deposited into a Curve LDO/wstETH TwoCrypto pool; LP tokens are retained by the LiquidityProvisioner as DAO-owned liquidity.

Reviewer: ```suggestion
1. **LP mode** — acquired LDO is paired with wstETH and deposited into a Curve LDO/wstETH TwoCrypto-NG pool; LP tokens are retained by the LiquidityProvisioner as DAO-owned liquidity.
```

Reviewer 2: fixed

2. **Treasury-only mode** — acquired LDO is delivered directly to the Aragon Agent treasury.

> The active mode is set via the `liquidityProvisioner` address on the NESTController: a non-zero address enables LP mode; `address(0)` enables treasury-only mode. Switching between modes requires a full DAO vote.

Reviewer: In this LIP treasury-only mode delivers LDO to the AGENT "directly" via Stonks receiver. 

Same time `Stonks v2` says "If [receiver] set to `address(0)`, it defaults to `AGENT`" 

The CoW protocol layer interprets `receiver = address(0)` as "settle to owner" (the order owner / Stonks contract itself). Thus, the LIP's wording suggests Stonks v2 will rewrite `address(0)` to AGENT before passing into `GPv2Order.Data.receiver`, so the protocol-layer ambiguity does not bite. 

This needs an explicit unit/integration test suite 🤔

Reviewer 3: 1. Address 0 in the LiquidityProvisioner context doesn't directly point to the Stonks receiver; it switches operational mode. 

2. We'll need a second Stonks instance with an Agent as a receiver; this is explicitly stated in the spec, and as part of the NEST deployment, we'll be deploying this Stonks instance as well. Added this info to LIP in the stonks v2 section.

This LIP proposes deploying NEST in **LP mode as the initial operating configuration**. Treasury-only mode is available as a governance-controlled fallback that could be enabled by a subsequent DAO vote without redeploying the core contracts.

### New Contracts

#### `NESTController`

> Central coordinator of the system. Holds an array of revenue source addresses, aggregates revenue from active sources on each trigger, evaluates eligibility gates, enforces daily and annual caps, and creates CoW Swap orders via Stonks v2. In LP mode, wraps half of the stETH budget to wstETH and transfers it to the LiquidityProvisioner. Exposes pass-through functions for Stonks and Order management accessible by the Treasury Management Committee (TMC) and Emergency Committee.

Reviewer: not sure if it should always wrap exactly half, but conceptually prob okay-ish

#### `RevenueSource` (abstract)

> Common interface for revenue tracking implementations. Each source stores the latest normalized daily revenue in USD, the report timestamp, and an immutable staleness window set at deployment. Exposes `getRevenue()` returning `(uint256 revenueUSD, uint256 reportTimestamp, bool isStale)`.

Reviewer: staleness window is not provided in the params list

Reviewer 2: fixed

Revenue from sources covering different time periods is normalized to a daily USD rate before storage:

```

> dailyRevenueUSD = revenueUSD × 86400 / (reportTimestamp − lastReportTimestamp)

Reviewer: near-to-zero denominator might inflate `dailyRevenueUSD` 🤔

Reviewer 2: This is a valid concern. However this will happen only in very temporally close reports with a non-zero revenue delta (two rebases with a delta firing almost at the same time) which is unlikely. Here we assume that the revenue sources in use all have sane reporting cadence. Further down the stream the inflation is ultimately bounded by the daily spend gate. 
I think that this can be hardened by enforcing the reports frequency with a limiter in each further source that we define, if it needs it (like in intra-day reporters that we do not have atm).

```

This ensures values from different reporting cadences are directly comparable. The controller sums normalized values from all active sources without per-source period tracking.

> #### `StakingRevenueSource`

Reviewer: probably would add a more detailed formula here, but not a blocker

If the rate does not increase, the source records zero revenue without updating the rate baseline. After a slashing event, the rate baseline remains at its pre-slash level and revenue stays zero until the protocol fully recovers past the previous high-water mark, ensuring a buyback pause proportional to slash severity.

Access to `pushTokenRate` is restricted to `REPORTER_ROLE` holders (granted to `TokenRateNotifier` at deployment).

`StakingRevenueSource` is the only revenue source integrated at launch. The modular `RevenueSource` interface and the controller's source array provide the extension path for future revenue streams. Each new source requires a dedicated contract, a governance vote for deployment and registration, and appropriate staleness configuration.

#### `LiquidityProvisioner`

Receives settled LDO directly from CoW Swap and wstETH from the NESTController. Deposits balanced amounts of both into the Curve LDO/wstETH pool. Retains minted LP tokens with managed withdrawal and transfer capabilities. Handles excess wstETH cleanup via a permissionless cooldown-gated unwrap function.

### Modified Existing Contracts

> #### `Stonks v2`

Reviewer: would need to give a reference to stonks v2 in the doc iteself

Reviewer: I mean there is a bit of mystery now what's Stonks v2 for the reader here

Reviewer 2: fixed

4. **Creates order** — sends stETH to Stonks v2 for LDO purchase via CoW Swap.
5. **Wraps and transfers wstETH** _(LP mode only)_ — wraps the other half of stETH to wstETH and transfers it to the LiquidityProvisioner.
6. **Updates state** — `lastTriggerOrderTimestamp` is set only when an order is actually created.

The controller uses two independent timestamps to prevent DoS via accounting-trigger griefing:

- `lastAccountingTimestamp` gates accounting cadence; prevents double-counting.
- `lastTriggerOrderTimestamp` gates order creation; prevents multiple orders per accounting cycle.

If accounting runs but an eligibility gate blocks the order, the keeper can call again later that day. Accounting will be skipped on retry, but order creation remains available.

#### Eligibility Gates

**Infrastructure gates (revert on failure):** execution pause, revenue sources registered, accounting cadence, order-already-created-in-cycle.

**Eligibility gates (non-reverting; emit `ExecutionSkipped` + return `address(0)`):** surplus positive, cumulative capacity non-negative, oracle quotability, ETH price floor, annual cap, stETH balance, minimum order size.

#### Surplus Mechanics

> Revenue surplus accumulates cumulatively. Each day's surplus increases the total buyback allocation; each day's loss reduces it. Buybacks are only allowed while cumulative unrealized capacity remains non-negative.

Reviewer: does it have negative floor though?

Reviewer 2: hmm, what do you mean by the negative floor?

Reviewer: to what extent it can go down below zero

Reviewer 2: gotcha, no, it doesn't. On one hand adding it will introduce a limit on how low we can go, on the other might be challenging to get the parameters right and we already have a remedy for such conditions. We have a dedicated governance action to trigger the `resetBuybackAccounting`. This will reset the current accumulators and allow buybacks to continue even if the conditions otherwise programmatically shouldn't (very deep slashes, prolonged unfavourable conditions where the deficit accumulated too much).

```
surplus = totalDailyRevenue − dailyRevenueThresholdUSD
dailyAllocation = surplus × surplusShareBps / 10000
allocatedForBuybacksUSD += dailyAllocation
unrealizedBuybacksUSD = allocatedForBuybacksUSD − cumulativeBuybacksUSD
```

Execution requires all of the following:

1. `lastDailyAllocationUSD > 0` implies today's surplus is positive.
2. `unrealizedBuybacksUSD ≥ 0` before today's allocation update — the protocol has not already spent more than it has earned.

> 3. Oracle quotability.

Reviewer: staleness check?

Reviewer 2: fixed

4. `ethPriceUSD > ethPriceFloorUSD` — inactive when floor is `0`.

Budget per trigger: `min(dailyCapUSD, lastDailyAllocationUSD)`, then clamped by annual cap and stETH balance. In LP mode, this constrained budget is split equally: one half is sold for LDO via Stonks, the other half is wrapped to wstETH and transferred to the LiquidityProvisioner. In treasury-only mode, the full budget is swapped for LDO.

**Transition behaviors:**

- **Surplus to deficit**: daily allocations go negative; `unrealizedBuybacksUSD` turns negative; buybacks pause until new surplus covers the deficit.
- **Deficit to surplus**: positive allocations gradually rebuild `allocatedForBuybacksUSD`; buybacks resume once `unrealizedBuybacksUSD` crosses back above zero.
- **Share changes**: affect only future allocations; past committed allocations are non-retroactive.

---

> ### Liquidity Provisioning

Reviewer: bootstrap path for `addLiquidity()` is missing:

If the Curve pool starts at zero, who creates the pool, with what initial price, A/gamma/fee/MA params, admin settings, and first deposit? What happens before `price_oracle()` has a meaningful EMA? 

IMO the LIP should define a bootstrap phase separate from steady-state LP mode.

Reviewer 3: 1. Pool deployment is an operational step executed by Lido Labs contributors. Initial price, A/gamma/fee/MA params, and admin settings are deployment-time decisions made by the contributors based on market conditions. Entry liquidity is funded from the EGG grant allocation, a small seed to initialize the pool before NEST orders begin settling. 
 
 2. Regarding EMA, we'll address it. The exact calibration requires analysis, rather than defining a formal bootstrap phase in the LIP. Do you think that's a blocker, or could it be a disclaimer that this part will be addressed and decided before the deployment?

The LiquidityProvisioner receives settled LDO from CoW Swap and wstETH from the NESTController. Liquidity provisioning runs as a separate permissionless flow after trade settlement.

`addLiquidity()` flow:

1. Read LDO and wstETH balances; revert if either is zero.
2. Convert both to USD via the OracleRouter; revert if any price feed is stale or unavailable.
3. Take the minimum of the two USD values and compute balanced token amounts corresponding to that minimum.
4. Deposit LDO and wstETH into the Curve pool with a minimum mint amount slippage guard.
5. Retain minted LP tokens in the provisioner.

Excess tokens on the larger side remain in the provisioner for the next `addLiquidity()` cycle. The Curve LDO/wstETH pool starts with zero balance and scales naturally as orders settle — no upfront seeding is required.

#### Pool Skew Protection

Two-layer protection scheme:

> 1. **Pool price divergence gate** compares the Curve pool's internal EMA price from `price_oracle()` against the OracleRouter LDO/wstETH price. If divergence exceeds `poolPriceDivergenceToleranceBps`, the deposit reverts. This is a structural check independent of deposit size.

Reviewer: Comparing a low-liquidity pool EMA to OracleRouter price may reject all deposits after normal market moves, while lack of trades prevents EMA convergence. 

Maybe define bootstrap bypass, minimum liquidity, minimum oracle age, and recovery controls.

2. **LP token slippage guard** sets a minimum mint amount from `calc_token_amount()` discounted by `poolSlippageToleranceBps`, guarding against MEV sandwich attacks during the deposit transaction.

#### Excess wstETH

Each trigger in LP mode wraps and transfers wstETH equal to the sell half. Partial or unfilled orders leave excess wstETH in the provisioner with no LDO counterpart. There are two resolution paths:

- **`retryFromStonks()`** recovers stETH from an expired Order to Stonks and creates a new order. Shares ETH price floor and oracle quotability gates with `triggerExecution()` but bypasses budget gates since stETH was already committed.

> - **`unwrapExcessWstEth()`** unwraps excess wstETH to stETH and returns it to the NESTController after a cooldown. Uses a clamp target accounting for all sell-side stETH still in the pipeline. After transfer, `accountForReturnedExcess()` on the controller adjusts both `annualSpendAccumulatorUSD` and `cumulativeBuybacksUSD` to prevent double-counting.

Reviewer: cooldown is not specified below

Reviewer 2: fixed directly in text

#### LP Token Management

| Function                                         | Access               | Description                                                          |
| ------------------------------------------------ | -------------------- | -------------------------------------------------------------------- |
| `removeLiquidity(uint256 lpAmount)`              | `MANAGER_ROLE`       | Burns LP tokens, withdraws LDO and wstETH back to provisioner        |
| `recoverERC20(address token, uint256 amount)`    | `MANAGER_ROLE`       | Recovers assets to the Aragon Agent treasury                         |
| `transferLpTokensTo(address to, uint256 amount)` | `DEFAULT_ADMIN_ROLE` | Transfers LP tokens to an arbitrary address — requires full DAO vote |

> 

Author: Table looks fine

| **Gate Thresholds** | `ethPriceFloorUSD`                | ETH price eligibility condition. `0` disables the gate.                                                                                                                                                                     |
| **Revenue**         | `dailyRevenueThresholdUSD`        | Daily revenue floor for surplus computation                                                                                                                                                                                 |
|                     | `surplusShareBps`                 | Share of daily surplus allocated to NEST per trigger (basis points). In LP mode, this allocation is split equally between the LDO purchase leg and the wstETH pairing leg.                                                  |
| **Spending Cap**    | `dailyCapUSD`                     | USD limit per trigger. Must be strictly less than `annualCapUSD`.                                                                                                                                                           |

> | **Annual Cap**      | `annualCapUSD`                    | USD limit on spend committed within a fixed 365-day period that resets when elapsed. Returned stETH from expired or unfilled orders is credited back via `accountForReturnedExcess()`, partially restoring annual headroom. |

Reviewer: should it be fixed or rolling period tho?

Reviewer 2: at the design stage after discussion we chose the fixed period for the annual cap

Reviewer 3: We considered both options and chose the fix, rolling requires per-transaction timestamp tracking to age out spend. Fixed tracks one counter against one start timestamp — simpler state, simpler audit, identical safety.

| **Pool Guards**     | `poolSlippageToleranceBps`        | Slippage guard on Curve deposits (MEV protection)                                                                                                                                                                           |
|                     | `poolPriceDivergenceToleranceBps` | Divergence guard against pool skew                                                                                                                                                                                          |
| **Revenue Sources** | `revenueSources`                  | Array of active revenue source addresses                                                                                                                                                                                    |
| **Provisioner**     | `liquidityProvisioner`            | LiquidityProvisioner address. `address(0)` = treasury-only mode.                                                                                                                                                            |

> 

Author: Intentional: some params are unified under single category.

> All parameters are configurable via Aragon Voting (`DEFAULT_ADMIN_ROLE`) with validation on write. Invalid configuration reverts and preserves the previous valid state.

Reviewer: would probably mention that it's deliberately not under DG

Reviewer 2: added

| `DEFAULT_ADMIN_ROLE` + `MANAGER_ROLE` + `EMERGENCY_ROLE` | Aragon Voting                       | Full configuration control on both contracts, all emergency controls, asset recovery, role management. Requires full DAO vote.                                 |
| `MANAGER_ROLE` + `EMERGENCY_ROLE`                        | Treasury Management Committee (TMC) | Asset recovery, LP token management (`removeLiquidity`, `recoverERC20`), emergency pause/unpause, propose EasyTrack motions to fund the controller with stETH. |
| `EMERGENCY_ROLE`                                         | Emergency Committee / Multisig      | Pause controls on all NEST contracts, emergency order operations (cancel, revoke relayer). Cannot modify configuration.                                        |
| —                                                        | Permissionless Callers              | `triggerExecution()`, `retryFromStonks()`, `addLiquidity()`, `unwrapExcessWstEth()`                                                                            |

> 

Author: Intentional: one cell is empty to highlight permissionless actions.

2. **LiquidityProvisioner liquidity pause** — stops `addLiquidity()`. `unwrapExcessWstEth()` remains callable.
3. **Revenue source pauses** — stops data acceptance. Pausing all sources halts execution via `NoActiveRevenueSources` revert without engaging the execution pause.
4. **Stonks creation/signature pauses** — stops new order creation or freezes settlement of existing orders independently.
5. **OracleRouter pause** — stops price queries, causing trade execution and liquidity provisioning to revert.

Asset recovery is available at all times and is not subject to pause. NESTController and LiquidityProvisioner auto-unwrap wstETH to stETH before transferring to treasury.

**Accounting reset:** If cumulative buyback accounting becomes untenable — for example, after prolonged slashing recovery where accumulated deficit would block buybacks indefinitely — governance can call `resetBuybackAccounting` to zero out `allocatedForBuybacksUSD`, `cumulativeBuybacksUSD`, and `lastDailyAllocationUSD`. The next accounting cycle starts fresh. Requires a full DAO vote (`DEFAULT_ADMIN_ROLE`).

---

> ## Rationale

Reviewer: would expect rationale on pool model selection tbh (IL, position management, etc.)

### $50,000 Daily Cap as Primary Safety Limiter

The daily cap is the primary active safety guardrail. Historically, very few days produce surplus exceeding $50k. In the event of oracle corruption or bad revenue inputs, the cap ensures maximum exposure over a ~6-day governance response window is bounded at ~$300k. The cap is independently configurable and can be raised by governance as surplus and on-chain liquidity grow.

### Zero-seed Pool Launch

No upfront LDO seeding from the Aragon Agent treasury is required. Seeding the pool from treasury while simultaneously buying LDO from the market is logically contradictory: the two operations work against each other, with treasury supply injection partially offsetting the market buy. There is no mechanical benefit to pre-seeding.

> Additionally, a smaller pool is more resistant to value extraction. A smaller pool has a steeper bonding curve, so price gaps created by buyback activity close faster, leaving less time for arbitrageurs to capture the spread. Starting at zero and growing organically through buyback activity is both the simpler and the more arbitrage-resistant launch posture.

Reviewer: > smaller pool is more resistant

I think it's quite opposite tbh, so would reconsider this argument or remove it

Reviewer 3: What is your reason to think the opposite?

Reviewer 3: Adjusted

| ------------------------------------ | --------------------------------------------------------------------------------------- |
| Manual monthly buybacks              | Reintroduces discretionary human intervention; not permissionless; regulatory risk      |
| Quarterly revenue smoothing          | No meaningful improvement in execution volume; higher accounting and oracle complexity  |
| Instant daily-spend (non-cumulative) | No memory of losses; risk of overspending during consecutive surplus days               |
| Burn acquired LDO                    | One-time signal for exit-seeking traders, not a durable mechanism for long-term holders |
| Simple revenue percentage split      | Does not enforce surplus discipline; can activate with no actual treasury surplus       |

Expired and unfilled orders are retried through a bounded recovery flow (`retryFromStonks()`) rather than silently skipped. This is a resolved design decision: recovered stETH is reused for a new order without additional funding, while excess wstETH from partial fills is unwrapped and returned to the controller via a cooldown-gated permissionless function.

---

> ## Security Considerations

Reviewer: Pls add Missing Lido AccountingOracle report(s)

Reviewer 2: added

### Permissionless Execution Risk

`triggerExecution()` is callable by anyone. Budget computation uses on-chain oracle prices at call time, with `orderPriceProtectionBps` reducing `minBuyAmount` to absorb price movement between submission and inclusion. Daily and annual caps enforce hard spending bounds regardless of oracle precision.

### Revenue Reporting Integrity

`StakingRevenueSource` access to `pushTokenRate` is restricted to `REPORTER_ROLE` holders (granted to `TokenRateNotifier` at deployment). Out-of-order calls by unauthorized callers are rejected. Revenue sources can be individually paused.

### Slashing Events

After a slashing event, the revenue rate baseline stays at its pre-slash level. Staking revenue returns zero until the protocol fully recovers, creating a natural buyback pause proportional to slash severity. Governance can reset accounting if accumulated deficit becomes indefinitely blocking.

> ### Asset Custody

Reviewer: ```suggestion
### No asset custody
```

Reviewer 2: fixed

All purchased assets remain DAO-owned throughout the execution path. stETH flows only through the controlled pipeline: Aragon Agent → NESTController → Stonks → Order → LiquidityProvisioner (LP mode) or Aragon Agent (treasury-only mode). No unauthorized exit path exists for sell-side stETH.

### Emergency Response Time

Five independent pause domains allow targeted response without halting the full system. The Emergency Committee can pause without a full DAO vote. Asset recovery functions are always available and not subject to pause. At a $50k daily cap, the maximum exposure over a 6-day governance response window is ~$300k.

---

> ## References

Reviewer: would either add NEST repo here or have a separate `## Test cases` with this ref to impl and tests

Reviewer 2: fixed

- Forum discussion: [Liquid Buybacks: NEST Execution with LDO/wstETH Liquidity](https://research.lido.fi/t/liquid-buybacks-nest-execution-with-ldo-wsteth-liquidity/10894)

> - [Stonks v2](https://github.com/lidofinance/stonks)

Reviewer: would rather refer to a doc bc stonks v1 also lived here and repository readme doesn't help a lot there

- [CoW Protocol documentation](https://docs.cow.fi/)

> - [Curve TwoCrypto](https://docs.curve.fi/)

Reviewer: not sure if it is the good link
https://github.com/curvefi/twocrypto-ng ?

Reviewer 2: yeah, it's better. Updated

> - [TokenRateNotifier](https://github.com/lidofinance/lido-dao)

Reviewer: can we use https://github.com/lidofinance/core/blob/master/contracts/0.8.9/TokenRateNotifier.sol ?

Reviewer 2: fixed

`addLiquidity()` flow:

1. Read LDO and wstETH balances; revert if either is zero.
2. Convert both to USD via the OracleRouter; revert if any price feed is stale or unavailable.
3. Take the minimum of the two USD values and compute balanced token amounts corresponding to that minimum.
4. Deposit LDO and wstETH into the Curve pool with a minimum mint amount slippage guard.
5. Retain minted LP tokens in the provisioner.

Excess tokens on the larger side remain in the provisioner for the next `addLiquidity()` cycle. The Curve LDO/wstETH pool starts with zero balance and scales naturally as orders settle — no upfront seeding is required.

#### Pool Deployment

Deployment parameters, including initial price, A/gamma coefficients, fee structure, and oracle settings, will be published on the forum one week before deployment. Curve pools have no admin or pause controls post-deployment and are immutable once live. Emergency controls are scoped to the NEST contracts (see Emergency Controls).

#### Pool Bootstrap

> The pool launches with zero liquidity. Applying the divergence check before the pool has reached sufficient depth would block addLiquidity() at launch; the check is bypassed during bootstrap and engages once the pool crosses a defined depth threshold. The threshold value will be published on the forum two weeks before the on-chain vote.

Reviewer: why not two weeks before for both?

Reviewer 2: updated to two weeks
